File: compatibility/pg/csharp/PGTest.cs
using System;
using System.Collections.Generic;
using System.Data;
using Npgsql;
using System.IO;

public class PGTest
{
    public class Tests
    {
        private NpgsqlConnection conn;
        private NpgsqlCommand cmd;
        private List<Test> tests = new List<Test>();

        public void Connect(string ip, int port, string user, string password)
        {
            string connectionString = $"Host={ip};Port={port};Username={user};Password={password};Database=postgres;Timeout=300;CommandTimeout=600;";
            try
            {
                conn = new NpgsqlConnection(connectionString);
                conn.Open();
                cmd = conn.CreateCommand();
                cmd.CommandType = CommandType.Text;
            }
            catch (NpgsqlException e)
            {
                throw new Exception($"Error connecting to database: {e.Message}", e);
            }
        }

        public void Disconnect()
        {
            try
            {
                cmd.Dispose();
                conn.Close();
            }
            catch (NpgsqlException e)
            {
                throw new Exception(e.Message);
            }
        }

        public void AddTest(string query, string[][] expectedResults)
        {
            tests.Add(new Test(query, expectedResults));
        }

        public bool RunTests()
        {
            foreach (var test in tests)
            {
                if (!test.Run(cmd))
                {
                    return false;
                }
            }
            return true;
        }

        public void ReadTestsFromFile(string filename)
        {
            try
            {
                using (var reader = new StreamReader(filename))
                {
                    string line;
                    while ((line = reader.ReadLine()) != null)
                    {
                        if (string.IsNullOrWhiteSpace(line)) continue;
                        string query = line;
                        var results = new List<string[]>();
                        while ((line = reader.ReadLine()) != null && !string.IsNullOrWhiteSpace(line))
                        {
                            results.Add(line.Split(','));
                        }
                        string[][] expectedResults = results.ToArray();
                        AddTest(query, expectedResults);
                    }
                }
            }
            catch (IOException e)
            {
                Console.Error.WriteLine(e.Message);
                Environment.Exit(1);
            }
        }

        public class Test
        {
            private string query;
            private string[][] expectedResults;

            public Test(string query, string[][] expectedResults)
            {
                this.query = query;
                this.expectedResults = expectedResults;
            }

            public bool Run(NpgsqlCommand cmd)
            {
                try
                {
                    Console.WriteLine("Running test: " + query);
                    cmd.CommandText = query;
                    using (var reader = cmd.ExecuteReader())
                    {
                        if (!reader.HasRows)
                        {
                            if (expectedResults.Length != 0)
                            {
                                Console.Error.WriteLine($"Expected {expectedResults.Length} rows, got 0");
                                return false;
                            }
                            Console.WriteLine("Returns 0 rows");
                            return true;
                        }
                        if (reader.FieldCount != expectedResults[0].Length)
                        {
                            Console.Error.WriteLine($"Expected {expectedResults[0].Length} columns, got {reader.FieldCount}");
                            return false;
                        }
                        int rows = 0;
                        while (reader.Read())
                        {
                            for (int col = 0; col < expectedResults[rows].Length; col++)
                            {
                                string result = reader.GetValue(col).ToString().Trim();
                                if (expectedResults[rows][col] != result)
                                {
                                    Console.Error.WriteLine($"Expected:\n'{expectedResults[rows][col]}'");
                                    Console.Error.WriteLine($"Result:\n'{result}'");
                                    return false;
                                }
                            }
                            rows++;
                        }
                        Console.WriteLine("Returns " + rows + " rows");
                        if (rows != expectedResults.Length)
                        {
                            Console.Error.WriteLine($"Expected {expectedResults.Length} rows");
                            return false;
                        }
                        return true;
                    }
                }
                catch (NpgsqlException e)
                {
                    Console.Error.WriteLine(e.Message);
                    return false;
                }
            }
        }
    }

    public static void Main(string[] args)
    {
        if (args.Length < 5)
        {
            Console.Error.WriteLine("Usage: PGTest <ip> <port> <user> <password> <testFile>");
            Environment.Exit(1);
        }

        var tests = new Tests();
        tests.Connect(args[0], int.Parse(args[1]), args[2], args[3]);
        tests.ReadTestsFromFile(args[4]);

        if (!tests.RunTests())
        {
            tests.Disconnect();
            Environment.Exit(1);
        }
        tests.Disconnect();
    }
}

File: compatibility/pg/rust/pg_test.rs
use std::fs::File;
use std::io::{self, BufRead, BufReader};
use std::process::exit;

extern crate postgres;
use postgres::{Client, NoTls, Error};

struct Test {
    query: String,
    expected_results: Vec<Vec<String>>,
}

impl Test {
    fn new(query: String, expected_results: Vec<Vec<String>>) -> Self {
        Test { query, expected_results }
    }

    fn run(&self, client: &mut Client) -> bool {
        println!("Running test: {}", self.query);
        match client.query(&self.query, &[]) {
            Ok(rows) => {
                if rows.is_empty() {
                    if self.expected_results.is_empty() {
                        println!("Returns 0 rows");
                        return true;
                    }
                    eprintln!("Expected {} rows, got 0", self.expected_results.len());
                    return false;
                }
                if rows[0].columns().len() != self.expected_results[0].len() {
                    eprintln!("Expected {} columns, got {}", self.expected_results[0].len(), rows[0].columns().len());
                    return false;
                }
                for (i, row) in rows.iter().enumerate() {
                    for (j, expected) in self.expected_results[i].iter().enumerate() {
                        let result: String = row.get(j);
                        if expected != &result {
                            eprintln!("Expected:\n'{}'", expected);
                            eprintln!("Result:\n'{}'\nRest of the results:", result);
                            for row in rows.iter().skip(i + 1) {
                                eprintln!("{}", row.get::<usize, String>(0));
                            }
                            return false;
                        }
                    }
                }
                println!("Returns {} rows", rows.len());
                if rows.len() != self.expected_results.len() {
                    eprintln!("Expected {} rows", self.expected_results.len());
                    return false;
                }
                true
            }
            Err(err) => {
                eprintln!("{}", err);
                false
            }
        }
    }
}

struct Tests {
    client: Client,
    tests: Vec<Test>,
}

impl Tests {
    fn new(ip: &str, port: u16, user: &str, password: &str) -> Result<Self, Error> {
        let conn_str = format!("host={} port={} user={} password={} dbname=postgres", ip, port, user, password);
        let client = Client::connect(&conn_str, NoTls)?;
        Ok(Tests { client, tests: Vec::new() })
    }

    fn add_test(&mut self, query: String, expected_results: Vec<Vec<String>>) {
        self.tests.push(Test::new(query, expected_results));
    }

    fn run_tests(&mut self) -> bool {
        for test in &self.tests {
            if !test.run(&mut self.client) {
                return false;
            }
        }
        true
    }

    fn read_tests_from_file(&mut self, filename: &str) -> io::Result<()> {
        let file = File::open(filename)?;
        let reader = BufReader::new(file);
        let mut lines = reader.lines();
        while let Some(Ok(line)) = lines.next() {
            if line.trim().is_empty() {
                continue;
            }
            let query = line;
            let mut results = Vec::new();
            while let Some(Ok(line)) = lines.next() {
                if line.trim().is_empty() {
                    break;
                }
                results.push(line.split(',').map(String::from).collect());
            }
            self.add_test(query, results);
        }
        Ok(())
    }
}

fn main() {
    let args: Vec<String> = std::env::args().collect();
    if args.len() < 6 {
        eprintln!("Usage: {} <ip> <port> <user> <password> <testFile>", args[0]);
        exit(1);
    }

    let ip = &args[1];
    let port: u16 = args[2].parse().expect("Invalid port number");
    let user = &args[3];
    let password = &args[4];
    let test_file = &args[5];

    let mut tests = Tests::new(ip, port, user, password).expect("Failed to connect to database");
    tests.read_tests_from_file(test_file).expect("Failed to read test file");

    if !tests.run_tests() {
        exit(1);
    }
}